docker/registry/internal/tags.go

```diff
@@ -47,39 +48,55 @@ func getRepositoryOnly(s string) string {
 	return s[i+1:]
 }
 
+type tagFetcher interface {
+	url(string, ...interface{}) string
```

Comment:
although the unexported interface methods here look weird, I did check and this is used in a few other places in juju (esp. in the state package). Not sure if this is best practice but can always change in the future.

Reply:
This tagFetcher  interface was created because the fetchTag function could take baseClient or any other provider API client as an argument. And we don't want to expose these methods because they are only used by tagFetch function internally.